EIPS/eip-2935.md

```diff
@@ -52,25 +53,94 @@ def process_block_hash_history(block: Block, state: State):
             state.insert_slot(HISTORY_STORAGE_ADDRESS, ancestor.number % HISTORY_SERVE_WINDOW, ancestor.hash)
 ```
 
-Note that if this is the fork block, then it persists the additional requisite history that could be needed while resolving `BLOCKHASH` opcode for all of the `HISTORY_SERVE_WINDOW` ancestors (up until genesis).
+Note that if this is the fork block, then it persists the additional requisite history that could be needed while resolving `BLOCKHASH` opcode for all of the `HISTORY_SERVE_WINDOW` > `BLOCKHASH_OLD_WINDOW` ancestors (up until genesis).
 
 For resolving the `BLOCKHASH` opcode this fork onwards (`block.timestamp >= FORK_TIMESTAMP`), switch the logic to:
 
 ```python
 def resolve_blockhash(block: Block, state: State, arg: uint64):
   # check the wrap around range
-  if arg >= block.number or arg < max(block.number - HISTORY_SERVE_WINDOW, 0)
+  if arg >= block.number or (arg + HISTORY_SERVE_WINDOW) < block.number
     return 0
 
   return state.load_slot(HISTORY_STORAGE_ADDRESS, arg % HISTORY_SERVE_WINDOW)
 ```
 
+Exact evm assembly that can be used for the blockhash contract:
+
+```
+// check for <=32 (<33) byte size or else revert
+calldatasize
+push1 0x21
+gt
+push1 0x0a
+// jump and skip revert
+jumpi
+
+push0
+push0
+revert
+
+// check if  blocknumber > input or else return 0
+jumpdest
+push0
+calldataload
+number
+gt
+push1 0x1a
+jumpi
+
+// return 0
+push0
+push0
+mstore
+push1 0x20
+push0
+return
+
+// check if input + 8125 > blocknumber (input >= blocknumber - 8124) or else return 0
```

Comment:
I don't see where these two values come from?

Reply:
sorry its 8193 and 8192 respectively. fixed